### PR TITLE
Java frontend: add support for lambdas that require generating boxing, unboxing, widening casts and upcasts

### PR DESCRIPTION
**Pull request by smowton**
The Java lambda metafactory, per https://docs.oracle.com/javase/8/docs/api/java/lang/invoke/LambdaMetafactory.html, can require adapting the type of the referenced method to that of the target functional interface by introducing boxing, unboxing, primitive type widening and pointer upcasting conversions. This adds tests for a wide variety of such cases and implements the required suppport.

### CODE REVIEW

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 1]
```diff
@@ -38,6 +38,72 @@ bool is_java_string_type(const struct_typet &struct_type)
...
+      {"java::java.lang.Short",
+       {"java::java.lang.Short.shortValue:()S", java_short_type()}},
+    };
+
+  auto found = type_info_by_name.find(type_name);
+  return found == type_info_by_name.end() ? nullptr : &found->second;
+}
+
+const java_primitive_type_infot *
+get_java_primitive_type_info(const typet &maybe_primitive_type)
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_utils.cpp:69]: :pick: `optionalt` rather than a pointer

File: jbmc/src/java_bytecode/lambda_synthesis.cpp [hunk 2]
```diff
@@ -610,6 +610,82 @@ static symbol_exprt instantiate_new_object(
...
+  const bool required_is_pointer = can_cast_type<pointer_typet>(required_type);
+
+  if(original_is_pointer == required_is_pointer)
+  {
+    if(original_is_pointer && original_type != required_type)
+      return typecast_exprt{expr, required_type};
+    else
+      return expr;
+  }
+
```
Comment (thk123) [inline jbmc/src/java_bytecode/lambda_synthesis.cpp]: What's this case for?

**Review by thk123 — APPROVED** [review]
:pick: Commit messages frequently exceed 72 (80?) chars causing github to wrap them, otherwise looks perfect to me

File: jbmc/src/java_bytecode/lambda_synthesis.cpp (hunk 2 above)
Comment (thk123) [reply jbmc/src/java_bytecode/lambda_synthesis.cpp]: I think the answer is byte to int and Derived -> Base?

Comment (smowton) [reply jbmc/src/java_bytecode/lambda_synthesis.cpp]: Gains a comment in a later commit (it's for pointer or integer type clashes without boxing, where plain casting always does the trick (forcing pointers and widening integers, so long as only valid Java casts are fed in in the first place)

**Pull request merged**